BFWControls/Modules/Layout/View/PileView.swift

```diff
@@ -3,57 +3,117 @@
+        return totalHeight + (subviewCount - 1) * gapHeight
     }
     
-    // MARK: - UIView
+    /// Boolean which indicates whether the subviews can be layed out horizontally.
+    private var isHorizontal: Bool {
+        return maxWidth >= subviewsTotalWidth
+    }
     
+    /// Returns intrinsic content size by laying out subviews either horizontally or vertically.
     open override var intrinsicContentSize: CGSize {
-        return CGSize(width: UIViewNoIntrinsicMetric,
-                      height: isHorizontal
-                        ? subviewHeight
-                        : subviewCount * subviewHeight + (subviewCount - 1) * gapHeight)
+        var width: CGFloat = 0
+        var height: CGFloat = 0
+        if isHorizontal {
+            width  = subviews.reduce(0) { (result, subview) in
+                return (result + subview.intrinsicContentSize.width)
+            }
+            width += (subviewCount - 1)*gapWidth
+            height  = subviews.reduce(0) { (result, subview) in
+                return max(result, subview.intrinsicContentSize.height)
+            }
+        } else  {
+            width  = subviews.reduce(0) { (result, subview) in
+                return max(result, subview.intrinsicContentSize.width)
+            }
+            height  = subviews.reduce(0) { (result, subview) in
+                return (result + subview.intrinsicContentSize.height)
+            }
+            height += (subviewCount - 1)*gapHeight
+        }
+        return CGSize(width: width, height: height)
+    }
+    
+    
+    /// Returns origin for a subview
+    ///
+    /// - Parameters:
+    ///   - subview: Current subview
+    ///   - previousSubview: Previous subview
+    /// - Returns: Origin as CGPoint
+    private func getOrigin(for subview: UIView, previousSubview: UIView?) -> CGPoint {
+        var subviewOriginY: CGFloat = 0.0
+        var subviewOriginX: CGFloat = 0.0
+        if isHorizontal {
+            subviewOriginX = previousSubview != nil
+                ? previousSubview!.frame.origin.x + previousSubview!.bounds.size.width + gapWidth
+                : 0.0
+        } else  {
+            subviewOriginY = previousSubview != nil
+                ? previousSubview!.frame.origin.y + previousSubview!.bounds.size.height + gapHeight
+                : 0.0
+        }
+        return CGPoint(x: subviewOriginX, y: subviewOriginY)
     }
     
+    /// Subviews layed out horizontally or vertically.
     open override func layoutSubviews() {
         invalidateIntrinsicContentSize()
         for (index, subview) in subviews.enumerated() {
-            let origin = isHorizontal
-                ? CGPoint(x: (subviewWidth + gapWidth) * CGFloat(index),
-                          y: 0.0)
-                : CGPoint(x: 0.0,
-                          y: (subviewHeight + gapHeight) * CGFloat(index))
+            for eachContraint in subview.constraints {
+                subview.removeConstraint(eachContraint)
+            }
```

Comment:
Can we replace the `for` loop with a single line functional `forEach`?:

```swift
subview.constraints.forEach { subview.removeConstraint($0) }
```

or:

```swift
subview.constraints.forEach(subview.removeConstraint)
```